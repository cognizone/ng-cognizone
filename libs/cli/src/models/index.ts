export * from './angular-json';
export * from './package-json';
