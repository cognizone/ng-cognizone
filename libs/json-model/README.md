# json-model

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test json-model` to execute the unit tests.
