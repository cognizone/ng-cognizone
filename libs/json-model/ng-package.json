{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/json-model",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
