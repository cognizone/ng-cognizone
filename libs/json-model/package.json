{
  "name": "@cognizone/json-model",
  "version": "3.7.1",
  "peerDependencies": {
    "@cognizone/lod": ">=3.0.0",
    "@cognizone/model-utils": ">=3.0.0",
    "@cognizone/ng-core": ">=3.0.0"
  },
  "dependencies": {
    "tslib": "^2.2.0"
  },
  "license": "Apache-2.0",
  "homepage": "https://github.com/cognizone/ng-cognizone",
  "contributors": [
    {
      "name": "William Chelman",
      "email": "william.chelman@cogni.zone"
    },
    {
      "name": "Farah El Agha",
      "email": "farah.agha111@gmail.com"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cognizone/ng-cognizone.git",
    "directory": "libs/json-model"
  },
  "private": false,
  "publishConfig": {
    "access": "public"
  }
}
