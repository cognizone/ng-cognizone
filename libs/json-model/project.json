{
  "name": "json-model",
  "projectType": "library",
  "sourceRoot": "libs/json-model/src",
  "prefix": "cz",
  "targets": {
    "build": {
      "executor": "@nrwl/angular:package",
      "outputs": ["{workspaceRoot}/dist/libs/json-model"],
      "options": {
        "project": "libs/json-model/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/json-model/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/json-model/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/json-model"],
      "options": {
        "jestConfig": "libs/json-model/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": ["libs/json-model/src/**/*.ts", "libs/json-model/src/**/*.html"]
      },
      "outputs": ["{options.outputFile}"]
    }
  },
  "tags": []
}
