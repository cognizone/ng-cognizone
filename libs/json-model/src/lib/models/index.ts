export * from './json-model';
export * from './resource';
