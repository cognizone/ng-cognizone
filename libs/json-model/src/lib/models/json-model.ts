/* eslint-disable @typescript-eslint/naming-convention */
import { Many, TypedResourceContext } from '@cognizone/model-utils';

/**
 * @experimental
 */
export type Uri<T extends JsonModel> = string & { '@@meta_type_placeholder@@'?: T };

export interface JsonModel {
  '@id': string;
  '@type': JsonModelType;
  '@context'?: TypedResourceContext;
  '@facets'?: {};
}

export type JsonModelType<T extends Many<string> = string> = T extends string ? Many<string> | T | [T, ...string[]] : Many<string>;

export function isJsonModel(o: unknown): o is JsonModel {
  return typeof o === 'object' && o != null && '@id' in o && '@type' in o;
}

export interface JsonModels<T extends JsonModel = JsonModel> {
  [uri: string]: T;
}

export type JsonModelFlat<T = JsonModel> = {
  [key in keyof T]: T[key] extends JsonModel
    ? string
    : T[key] extends JsonModel | undefined
    ? string | undefined
    : T[key] extends JsonModel[]
    ? string[]
    : T[key] extends JsonModel[] | undefined
    ? string[] | undefined
    : T[key];
};

export interface JsonModelFlatGraph<T extends JsonModelFlat = JsonModelFlat> {
  rootUri: string;
  context?: TypedResourceContext;
  facets?: {};
  models: {
    [uri: string]: T;
  };
}

export interface JsonModelValue<T> {
  '@value': T;
  '@type': string;
}

export function isJsonModelValue<T = unknown>(o: unknown): o is JsonModelValue<T> {
  return typeof o === 'object' && o != null && '@type' in o && '@value' in o;
}
