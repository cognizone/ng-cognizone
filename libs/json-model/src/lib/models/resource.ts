import { Many, TypedResourceContext } from '@cognizone/model-utils';

/**
 * @description intermediary model used internally to convert JsonModel to
 * TypedResourceGraph (and the other way around)
 */
export interface ResourceGraph {
  data: Resource;
  included: Resource[];
  facets?: {};
  context?: TypedResourceContext;
}

/**
 * @description intermediary model used internally to convert JsonModel to
 * TypedResourceGraph (and the other way around)
 */
export interface Resource<T extends object = {}> {
  uri: string;
  type: Many<string>;
  references: { [referenceKey: string]: Many<string> }; // value is one or multiple uris
  attributes: { [P in keyof T]: Many<ResourceAttribute<T[P]>> };
}

/**
 * @description intermediary model used internally to convert JsonModel to
 * TypedResourceGraph (and the other way around)
 */
export interface ResourceAttribute<T> {
  value: T;
  dataType: string;
}
