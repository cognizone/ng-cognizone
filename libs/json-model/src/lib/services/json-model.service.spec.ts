/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
import { ApplicationProfile } from '@cognizone/application-profile';
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries -- TODO fix this
import { NgApplicationProfileModule } from '@cognizone/ng-application-profile';
import { LoggerModule } from '@cognizone/ng-core';
import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';

import { JsonModelModule } from '../json-model.module';
import { JsonModel, JsonModelFlatGraph } from '../models/json-model';
import { JsonModelService } from './json-model.service';

const directiveAp: ApplicationProfile = require('../../test/directive-ap.json');
const treatyAp: ApplicationProfile = require('../../test/treaty-ap.json');
const treatyGraph1: JsonModelFlatGraph = require('../../test/treaty-process-flat-graph.json');

describe('JsonModelService', () => {
  const createService = createServiceFactory({
    service: JsonModelService,
    imports: [NgApplicationProfileModule.forRoot(), LoggerModule.forRoot('TEST'), JsonModelModule.forRoot()],
  });
  let spectator: SpectatorService<JsonModelService>;
  beforeEach(() => (spectator = createService()));

  const oldNow = Date.now.bind(Date);
  beforeAll(() => (Date.now = () => 0));
  afterAll(() => (Date.now = oldNow));

  it('should be created', () => {
    expect(spectator.service).toBeTruthy();
  });

  it('should transform to a JsonModelFlatGraph', () => {
    const root = getRoot();
    expect(spectator.service.toFlatGraph(root)).toMatchSnapshot();
  });

  it('should create a new JsonModel based on directive ap', () => {
    const json = spectator.service.createNewJsonModel(['EUDossier', 'Event'], directiveAp);
    expect(json).toMatchSnapshot();
  });

  it('should transform treaty flat graph to JsonModel', () => {
    const json = spectator.service.fromFlatGraph(treatyGraph1, treatyAp);
    expect(json).toMatchSnapshot();
  });

  it('should transform treaty flat graph to JsonModel and back', () => {
    const jsonModel = spectator.service.fromFlatGraph(treatyGraph1, treatyAp);
    const newGraph = spectator.service.toFlatGraph(jsonModel);
    expect(newGraph).toEqual(treatyGraph1);
  });
});

function getRoot(): JsonModel {
  const rootUri = 'uri1';
  const root = {
    '@id': rootUri,
    '@type': 'root',
    '@context': {
      base: 'http://example.org/',
      prefix: {
        xsd: 'http://www.w3.org/2001/XMLSchema#',
        rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
        rdfs: 'http://www.w3.org/2000/01/rdf-schema#',
      },
    },
    a: 4,
    children: ['just a uri, the is no JsonModel'],
    sibling: null,
    aNullStuff: null,
  };

  const sibling = {
    '@id': 'uri2',
    '@type': 'root',
    b: [4, 5, 6],
    sibling: root,
  };

  const child = {
    '@id': 'uri3',
    '@type': 'root',
    a: true,
    parent: root,
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (root as any).sibling = sibling;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  root.children.push(child as any);
  return root;
}
