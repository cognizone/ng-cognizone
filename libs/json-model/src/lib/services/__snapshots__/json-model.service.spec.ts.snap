// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonModelService should create a new JsonModel based on directive ap 1`] = `
Object {
  "@context": undefined,
  "@id": "http://resource/EUDossier-Event/0-0",
  "@type": Array [
    "EUDossier",
    "Event",
  ],
  "contentsEvent": Array [],
  "euDossierProducesResourceLegal": Array [],
  "euDossierResponsibilityOfAgent": Array [],
  "eventRiskLevel": undefined,
  "hasDocument": Array [],
  "hasEventRemark": Array [],
  "identifier": Array [],
  "initiatedByActPreparatory": Array [],
  "initiatesTransposition": undefined,
  "isAboutConceptTypeDossier": Array [],
  "relatedToToDo": Array [],
  "title": Object {},
}
`;

exports[`JsonModelService should transform to a JsonModelFlatGraph 1`] = `
Object {
  "context": Object {
    "base": "http://example.org/",
    "prefix": Object {
      "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
      "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
      "xsd": "http://www.w3.org/2001/XMLSchema#",
    },
  },
  "models": Object {
    "uri1": Object {
      "@id": "uri1",
      "@type": "root",
      "a": 4,
      "aNullStuff": null,
      "children": Array [
        "just a uri, the is no JsonModel",
        "uri3",
      ],
      "sibling": "uri2",
    },
    "uri2": Object {
      "@id": "uri2",
      "@type": "root",
      "b": Array [
        4,
        5,
        6,
      ],
      "sibling": "uri1",
    },
    "uri3": Object {
      "@id": "uri3",
      "@type": "root",
      "a": true,
      "parent": "uri1",
    },
  },
  "rootUri": "uri1",
}
`;

exports[`JsonModelService should transform treaty flat graph to JsonModel 1`] = `
Object {
  "@context": Object {
    "base": "http://example.org/",
    "prefix": Object {
      "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
      "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
      "xsd": "http://www.w3.org/2001/XMLSchema#",
    },
  },
  "@facets": Object {
    "someTitle": "I love titles",
  },
  "@id": "https://fedlex.data.admin.ch/treatyProcess/9999/5414",
  "@type": "TreatyProcess",
  "bilateral": false,
  "confidentialTreaty": false,
  "producedIn": Array [
    Object {
      "@id": "new-model-uri-initiation-1-1583078160543",
      "@type": "Initiation",
      "eventHasRemark": Array [],
      "factor": Object {
        "@id": "new-model-uri-internationalactor-0-1583078160540",
        "@type": "InternationalActor",
        "partyOrganisationOfTreaty": Array [
          Object {
            "@id": "new-model-uri-treatyprocess-2-1583078162434",
            "@type": "TreatyProcess",
            "committeeCreationBasedOn": Array [],
            "hasRemark": Array [],
            "institutionInChargeOfTheTreatyLevel2": Array [],
            "producedIn": Array [],
            "titleTreaty": Object {
              "de": Array [
                "",
              ],
              "fr": Array [
                "a",
              ],
              "it": Array [
                "",
              ],
            },
            "treatyAppliesEULegalResource": Array [],
            "treatyBasedOnTreaty": [Circular],
            "treatyComment": Array [],
            "treatyCommentAboutValidity": Array [],
            "treatyDescription": Array [],
            "treatyHasEntryIntoForceScope": Array [],
            "treatyHasRatificationRestriction": Array [],
            "treatyLegalResourceForOtherParty": Object {
              "@id": "new-model-uri-eulegalresource-4-1583078162435",
              "@type": "EULegalResource",
            },
            "treatyLimitDateForSignature": Array [],
            "treatyLimitDateForSignatureComment": Array [],
            "treatyNotificationPublishedIn": Array [],
            "treatyNotificationPublishedInLabel": Array [],
            "treatyPartyCountry": Array [],
            "treatyPartyOrganisation": [Circular],
            "treatyProcessHasResultingTreatyDocument": Array [],
            "treatyProcessHasTask": Array [],
            "treatyRelatedLegalResource": Array [],
            "treatySubject": Array [],
            "treatyType": Object {
              "@id": "https://fedlex.data.admin.ch/vocabulary/treaty-type/2",
              "@type": "TreatyType",
            },
          },
        ],
      },
    },
  ],
  "titleTreaty": Object {
    "de": Array [
      "Protokoll",
    ],
    "en": Array [
      "Protocol",
    ],
    "fr": Array [
      "Protocole",
    ],
    "it": Array [
      "Protocollo",
    ],
  },
  "treatyAppliesEULegalResource": Array [
    Object {
      "@id": "http://publications.europa.eu/resource/cellar/d67cf7eb-de13-4bcb-9959-0dc0107e102e",
      "@type": "EULegalResource",
    },
    Object {
      "@id": "http://publications.europa.eu/resource/cellar/87245951-547e-11ea-aece-01aa75ed71a1",
      "@type": "EULegalResource",
    },
  ],
  "treatyComment": Array [
    "This is a comment",
  ],
  "treatyProcessHasResultingTreatyDocument": Array [
    Object {
      "@id": "https://fedlex.data.admin.ch/treatyProcess/9999/5414/document/1",
      "@type": "TreatyDocument",
    },
    Object {
      "@id": "https://fedlex.data.admin.ch/treatyProcess/9999/5414/document/2",
      "@type": "TreatyDocument",
    },
  ],
  "treatySignatureDate": "2017-11-22T00:00:00.000Z",
  "treatySignaturePlace": "Strasbourg",
}
`;
